File: Glucloser/app/src/main/java/com/nlefler/glucloser/dataSource/ParseUploader.java
package com.nlefler.glucloser.dataSource;

import android.util.Log;

import com.nlefler.glucloser.models.BloodSugar;
import com.nlefler.glucloser.models.Meal;
import com.nlefler.glucloser.models.Place;
import com.parse.ParseException;
import com.parse.ParseObject;
import com.parse.SaveCallback;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import rx.Observable;
import rx.Observer;
import rx.Subscriber;
import rx.functions.Action1;
import rx.functions.Action2;
import rx.schedulers.Schedulers;

/**
 * Created by Nathan Lefler on 12/30/14.
 */
public class ParseUploader {
    private static String LOG_TAG = "ParseUploader";

    private static ParseUploader _sharedInstance;
    private Map<String, Observable<ParseObject>> inProgressUploads;

    public static synchronized ParseUploader SharedInstance() {
        if (_sharedInstance == null) {
            _sharedInstance = new ParseUploader();
        }
        return _sharedInstance;
    }
    private ParseUploader() {
        inProgressUploads = new HashMap<String, Observable<ParseObject>>();
    }

    public void uploadPlace(Place place) {
        final String placeId = place.getFoursquareId();
        this.getUploadedObjectObservable(placeId, place).subscribe(new Action1<ParseObject>() {
            @Override
            public void call(ParseObject parseObject) {
                parseObject.saveInBackground();
                inProgressUploads.remove(placeId);
            }
        });
    }

    public void uploadMeal(final Meal meal) {
        final Place place = meal.getPlace();
        final String placeId = place.getFoursquareId();

        final BloodSugar beforeSugar = meal.getBeforeSugar();
        final String beforeSugarId = beforeSugar != null ? beforeSugar.getId() : null;

        Observable<ParseObject> placeFetchObservable = getUploadedObjectObservable(placeId, place);
        Observable<ParseObject> beforeSugarObservable = null;
        if (beforeSugarId != null) {
            beforeSugarObservable = getUploadedObjectObservable(beforeSugarId, beforeSugar);
        }

        Observable<ParseObject> finalObservable = null;
        if (beforeSugarObservable != null) {
            placeFetchObservable.subscribeOn(Schedulers.io());
            finalObservable = Observable.merge(placeFetchObservable, beforeSugarObservable);
        } else {
            finalObservable = placeFetchObservable;
        }
        finalObservable.subscribe(new Observer<ParseObject>() {
            private ParseObject placeParseObject;
            private ParseObject beforeSugarParseObject;

            @Override
            public void onCompleted() {
                final String mealId = meal.getMealId();
                getUploadedObjectObservable(mealId, meal, placeParseObject, beforeSugarParseObject)
                        .subscribe(new Action1<ParseObject>() {
                            @Override
                            public void call(ParseObject mealObject) {
                                mealObject.saveInBackground();
                                inProgressUploads.remove(mealId);
                            }
                        });
            }

            @Override
            public void onError(Throwable e) {
                Log.e(LOG_TAG, "Unable to save Meal to Parse: " + e.getMessage());
            }

            @Override
            public void onNext(ParseObject parseObject) {
                String className = parseObject.getClassName();
                if (className.equals(Place.ParseClassName)) {
                    placeParseObject = parseObject;
                    inProgressUploads.remove(placeId);
                } else if (className.equals(BloodSugar.ParseClassName)) {
                    beforeSugarParseObject = parseObject;
                    inProgressUploads.remove(beforeSugarId);
                }
            }
        });
    }

    public void uploadBloodSugar(final BloodSugar sugar) {
        final String sugarId = sugar.getId();
        this.getUploadedObjectObservable(sugarId, sugar).subscribe(new Action1<ParseObject>() {
            @Override
            public void call(ParseObject parseObject) {
                parseObject.saveInBackground();
                inProgressUploads.remove(sugarId);
            }
        });
    }

    /** Helpers */
    private synchronized Observable<ParseObject> getUploadedObjectObservable(String localId,
                                                                             final Object toUpload,
                                                                             final Object... args) {
        if (this.inProgressUploads.containsKey(localId)) {
            return this.inProgressUploads.get(localId);
        } else {
            Observable<ParseObject> observable = Observable.create(new Observable.OnSubscribe<ParseObject>() {
                @Override
                public void call(final Subscriber<? super ParseObject> subscriber) {
                    if (toUpload == null) {
                        subscriber.onError(new IllegalArgumentException("Object to upload must not be null"));
                        return;
                    }
                    if (toUpload instanceof Place) {
                        PlaceFactory.ParseObjectFromPlace((Place)toUpload, createParseObjectReadyAction(subscriber));
                    } else if (toUpload instanceof Meal) {
                        if (args == null || args.length < 2 ||
                                !(args[0] instanceof ParseObject)) {
                            subscriber.onError(new IllegalArgumentException("Invalid specific arguments"));
                            return;
                        }
                        MealFactory.ParseObjectFromMeal((Meal)toUpload,
                                (ParseObject)args[0],
                                (ParseObject)args[1],
                                createParseObjectReadyAction(subscriber));
                    } else if (toUpload instanceof BloodSugar) {
                        BloodSugarFactory.ParseObjectFromBloodSugar((BloodSugar)toUpload,
                                createParseObjectReadyAction(subscriber));
                    } else {
                        subscriber.onError(new IllegalArgumentException("Invalid type for upload object"));
                    }
                }
            });
            this.inProgressUploads.put(localId, observable);
            return observable;
        }
    }

    private Action2<ParseObject, Boolean> createParseObjectReadyAction(final Subscriber<? super ParseObject> subscriber) {
        return new Action2<ParseObject, Boolean>() {
                            @Override
                            public void call(final ParseObject parseObject, Boolean created) {
                                if (parseObject == null) {
                                    subscriber.onError(new RuntimeException("Unable to create ParseObject"));
                                    return;
                                }

                                if (created) {
                                    parseObject.saveInBackground(new SaveCallback() {
                                        @Override
                                        public void done(ParseException e) {
                                            if (e == null) {
                                                subscriber.onNext(parseObject);
                                                subscriber.onCompleted();
                                            } else {
                                                subscriber.onError(e);
                                            }
                                        }
                                    });
                                } else {
                                    subscriber.onNext(parseObject);
                                    subscriber.onCompleted();
                                }
                            }
                        };
    }
}

File: Glucloser/app/src/main/java/com/nlefler/glucloser/dataSource/ParseUploader.kt
package com.nlefler.glucloser.dataSource

import android.util.Log
import com.nlefler.glucloser.models.*

import com.parse.ParseException
import com.parse.ParseObject
import com.parse.SaveCallback

import rx.Observable
import rx.Observer
import rx.Subscriber
import rx.functions.Action1
import rx.functions.Action2
import rx.schedulers.Schedulers
import java.util.*

/**
 * Created by Nathan Lefler on 12/30/14.
 */
public class ParseUploader private constructor() {
    private val inProgressUploads: MutableMap<String, Observable<ParseObject>>

    init {
        inProgressUploads = HashMap<String, Observable<ParseObject>>()
    }

    public fun uploadPlace(place: Place) {
        val placeId = place.getFoursquareId()!!
        this.getUploadedObjectObservable(placeId, place).subscribe({ parseObject: ParseObject ->
            parseObject.saveInBackground()
            inProgressUploads.remove(placeId)
        })
    }

    public fun uploadBolusEvent(bolusEvent: BolusEvent) {

        var finalObservable: Observable<ParseObject>? = null
        var placeId: String = ""
        if (bolusEvent is HasPlace) {
            val place = bolusEvent.getPlace()
            placeId = place.getFoursquareId()!!
            finalObservable = getUploadedObjectObservable(placeId, place)
        }

        val beforeSugar = bolusEvent.getBeforeSugar()
        val beforeSugarId = beforeSugar?.getId() ?: null
        if (beforeSugarId != null) {
//            finalObservable.subscribeOn(Schedulers.io())
            val bolusObservable = getUploadedObjectObservable(beforeSugarId, beforeSugar!!)
            if (finalObservable != null) {
                finalObservable = Observable.merge<ParseObject>(finalObservable, bolusObservable)
            }
            else {
                finalObservable = bolusObservable
            }
        }

        val foodIds = ArrayList<String>()
        for (food in bolusEvent.getFoods()) {
            val foodId = food.getFoodId()
            foodIds.add(foodId)
            finalObservable = Observable.merge<ParseObject>(finalObservable, getUploadedObjectObservable(foodId, food))
        }

        if (finalObservable == null) {
            Log.e(LOG_TAG, "Can't upload bolus event, final observable is null")
            return
        }

        finalObservable.subscribe(object : Observer<ParseObject> {
            var placeParseObject: ParseObject? = null
            var beforeSugarParseObject: ParseObject? = null
            val foodParseObjects = ArrayList<ParseObject>()

            override fun onCompleted() {
                val bolusEventId = bolusEvent.getId();
                getUploadedObjectObservable(bolusEventId, bolusEvent, beforeSugarParseObject, foodParseObjects, placeParseObject).subscribe({ mealObject: ParseObject ->
                    mealObject.saveInBackground()
                    inProgressUploads.remove(bolusEventId)
                })
            }

            override fun onError(e: Throwable) {
                Log.e(LOG_TAG, "Unable to save bolus event to Parse: " + e.getMessage())
            }

            override fun onNext(parseObject: ParseObject) {
                when (parseObject.getClassName()) {
                    Place.ParseClassName -> {
                        placeParseObject = parseObject
                        inProgressUploads.remove(placeId)
                    }
                    BloodSugar.ParseClassName -> {
                        beforeSugarParseObject = parseObject
                        inProgressUploads.remove(beforeSugarId)
                    }
                    Food.ParseClassName -> {
                        foodParseObjects.add(parseObject)
                        foodIds.remove(parseObject.get(Food.FoodIdFieldName))
                    }
                }
            }
        })
    }

    public fun uploadBloodSugar(sugar: BloodSugar) {
        val sugarId = sugar.getId()!!
        this.getUploadedObjectObservable(sugarId, sugar).subscribe({ parseObject: ParseObject ->
            parseObject.saveInBackground()
            inProgressUploads.remove(sugarId)
        })
    }

    public fun uploadFood(food: Food) {
        val foodId = food.getFoodId()
        this.getUploadedObjectObservable(foodId, food).subscribe({ parseObject: ParseObject ->
            parseObject.saveInBackground()
            inProgressUploads.remove(foodId)
        })
    }

    /** Helpers  */
    synchronized private fun getUploadedObjectObservable(localId: String, toUpload: Any, vararg args: Any?): Observable<ParseObject> {
        if (this.inProgressUploads.containsKey(localId)) {
            return this.inProgressUploads.get(localId)!!
        } else {
            val observable = Observable.create<ParseObject>(object : Observable.OnSubscribe<ParseObject> {
                override fun call(subscriber: Subscriber<in ParseObject>) {
                    when (toUpload) {
                        is Place -> {
                            PlaceFactory.ParseObjectFromPlace(toUpload, createParseObjectReadyAction(subscriber))
                        }
                        is Meal -> {
                            if (args.size() < 3 || !(args[0] is ParseObject && args[1] is ParseObject && args[2] != null)) {
                                subscriber.onError(IllegalArgumentException("Invalid specific arguments for Meal"))
                                return
                            }
                            @suppress("UNCHECKED_CAST")
                            MealFactory.ParseObjectFromMeal(toUpload,
                                    args[0] as ParseObject?,
                                    args[2] as ParseObject?,
                                    args[1] as List<ParseObject>,
                                    createParseObjectReadyAction(subscriber))
                        }
                        is BloodSugar -> {
                            BloodSugarFactory.ParseObjectFromBloodSugar(toUpload, createParseObjectReadyAction(subscriber))
                        }
                        is Snack -> {
                            if (args.size() < 2) {
                                subscriber.onError(IllegalArgumentException("Invalid specific arguments for Snack"))
                                return
                            }
                            @suppress("UNCHECKED_CAST")
                            SnackFactory.ParseObjectFromSnack(toUpload,
                                    args[0] as ParseObject?,
                                    args[1] as List<ParseObject>,
                                    createParseObjectReadyAction(subscriber))
                        }
                        is Food -> {
                            FoodFactory.ParseObjectFromFood(toUpload, createParseObjectReadyAction(subscriber))
                        }
                        else -> {
                            subscriber.onError(IllegalArgumentException("Invalid type for upload object"))
                        }
                    }
                }
            })
            this.inProgressUploads.put(localId, observable)
            return observable
        }
    }

    private fun createParseObjectReadyAction(subscriber: Subscriber<in ParseObject>): Action2<ParseObject?, Boolean> {
        return object : Action2<ParseObject?, Boolean> {
            override fun call(parseObject: ParseObject?, created: Boolean?) {
                if (parseObject == null) {
                    subscriber.onError(RuntimeException("Unable to create ParseObject"))
                    return
                }

                if (created!!) {
                    parseObject.saveInBackground({e: ParseException? ->
                        if (e == null) {
                            subscriber.onNext(parseObject)
                            subscriber.onCompleted()
                        } else {
                            subscriber.onError(e)
                        }
                    })
                } else {
                    subscriber.onNext(parseObject)
                    subscriber.onCompleted()
                }
            }
        }
    }

    companion object {
        private val LOG_TAG = "ParseUploader"

        private var _sharedInstance: ParseUploader? = null

        synchronized public fun SharedInstance(): ParseUploader {
            if (_sharedInstance == null) {
                _sharedInstance = ParseUploader()
            }
            return _sharedInstance!!
        }
    }
}
